packages/core/src/State.ts

```diff
@@ -288,57 +384,27 @@ export function createMachineSnapshot<
     tags: new Set(flatten(config.configuration.map((sn) => sn.tags))),
     children: config.children as any,
     historyValue: config.historyValue || {},
-    matches(parentStateValue) {
-      return matchesState(parentStateValue as any, this.value);
-    },
-    hasTag(tag) {
-      return this.tags.has(tag);
-    },
-    can(event) {
-      if (isDevelopment && !this.machine) {
-        console.warn(
-          `state.can(...) used outside of a machine-created State object; this will always return false.`
-        );
-      }
-
-      const transitionData = this.machine.getTransitionData(this as any, event);
+    // this one is generic in the target and it's hard to create a matching non-generic source signature
+    matches: machineSnapshotMatches as any,
+    hasTag: machineSnapshotHasTag,
+    can: machineSnapshotCan,
+    toJSON: machineSnapshotToJSON
+  };
 
-      return (
-        !!transitionData?.length &&
-        // Check that at least one transition is not forbidden
-        transitionData.some((t) => t.target !== undefined || t.actions.length)
-      );
+  Object.defineProperties(snapshot, {
```

Comment:
I'd really love to get rid of those getters. We don't even benefit from them being getters since we "tend to" spread those objects anyway, like here:
https://github.com/statelyai/xstate/blob/7c1497318799c6abe638e3dfa4639f86b880c18f/packages/core/src/State.ts#L415-L416

That evaluates them anyway, almost for each state. Currently we end up computing "a lot" of temporary states within a single transition and each `cloneMachineSnapshot` pays the price of invoking them. 

The problem isn't only unique to those getters. Some other things are computed by each `createMachineSnapshot` as well.

Reply:
What's the alternative? I think a `getNextTransitions()` is more accurate than a `nextEvents`, and we can have a `getMeta()` function.

Reply:
> What's the alternative?

I'd like to just replace those with methods or simple properties - this way we would end up with more predictable behavior around those and they just would work like everything else here. 

> I think a getNextTransitions() is more accurate than a nextEvents, and we can have a getMeta() function.

Ye, I think that making them methods like those would be good. That's a topic for a follow up PR though. I'm also not sure if `getNextTransitions` would be the best replacement name for `nextEvents` since we get a list of strings back and that doesn't scream "transitions" to me. Either way, we can bikeshed later on :p